### PR TITLE
Documentation and schema updates

### PR DESCRIPTION
**Pull request by GernotMaier**
This PR includes:

- significant improvement of main README.md
- documentation of schema for parameter and input data [schemas/README.md](schemas/README.md)
- simplified parameters schema as discussed in [gitlab pr 22](https://gitlab.cta-observatory.org/cta-computing/dpps/simulationpipeline/prototypes/notes/simulation-project-parameter-workflows/-/merge_requests/22). Note that the schema description is not 100% complete, but is thought to give a guidance on the approach.
- adaption of example schema files to this simplified description

### CODE REVIEW

File: schemas/MST_mirror_2f_measurements.schema.yml [hunk 1]
```diff
@@ -1,44 +1,39 @@
...
+        type: double
+      - name: psf
+        description: |-
+          Spot size of mirror panel PSF at nominal distance.
+        required: true
+        units: cm
+        type: double
+      - name: psf_opt
+        description: |-
+          Spot size of mirror panel PSF
```
Comment (VictorBarbosaMartins) [inline schemas/MST_mirror_2f_measurements.schema.yml:35]: please refer in the description also what "opt" means.

File: schemas/README.md [hunk 2]
```diff
@@ -1,13 +1,122 @@
...
+- `name`: Name of the parameter.
+- `description`: Description of the parameter.
+- `short_description`: Short description of the parameter.
+
+### Parameter data
+
+The *data* field is used to describe the actual type, format, and allowed values of the parameter.
+
+- `type`: Data type of the parameter.
+- `units`: Units of the parameter (compatible with astropy units). Note that units are explicitly listed in the jsonschema.yml file.
```
Comment (VictorBarbosaMartins) [inline schemas/README.md:72]: what does compatible with astropy units mean? Is it that if you pass it as an argument to`astropy.Units()` it will give you an instance of `astropy.Quantity` ? If yes, then please include this explanation.

File: schemas/README.md [hunk 3]
```diff
@@ -1,13 +1,122 @@
...
+        required_range:
+          min: 300.0
+          max: 700.0
+        input_processing:
+          - remove_duplicates
+          - sort
+      - name: photo_detection_efficiency
+        description: Average quantum or photon detection efficiency.
+        required: true
+        units: dimensionless
```
Comment (VictorBarbosaMartins) [inline schemas/README.md:97]: Related to the previous comment: I know that you prefer it to be more human readable, but note that if this is supposed to be astropy compatible, I think this should be`!astropy.units.Unit {unit: ''}`. (the same applies for all the units)

Comment (VictorBarbosaMartins) [reply schemas/README.md:97]: Ah, now I see that the allowed units are also listed in `schemas/jsonschema.yml`. Please mention that somewhere in the text as well.

**Review by VictorBarbosaMartins — CHANGES_REQUESTED** [review]
Hi @GernotMaier. Thanks for updating the documentation. Here I provide my comments, which consist of some suggestions to improve the text and mostly by questions, since I am not very used to the workflows. I will wait for some changes to approve, but this shouldnt take long I think. Since the your updates are only related to the documentation part I did not test the workflows. By the way, shouldnt we create tests for the workflows?
Thanks!

File: README.md [hunk 4]
```diff
@@ -2,53 +2,92 @@
 
 ## Introduction
 
-CWL implementations of simtools workflows:
+CWL implementations of workflows for the CTA Simulation Pipeline.
 
-- workflows for setting and validation of parameters are in [./workflows](./workflows).
-- command line tools used in these workflows and calling simtools applications are in [./workflows/tools](./workflows/tools).
-- input and generated data is validated using schemas located in [./schemas](./schemas).
+Workflows are used to set, derive, and validate simulation model parameters following all or a subset of these steps:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:7]: I don't think we want to use the cwl mechanism for the simulation production. Of course these are workflows, but we have dirac for that.

File: README.md [hunk 5]
```diff
@@ -2,53 +2,92 @@
...
 
-## Getting started
+Setting Workflows:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:9]: Wait - this is *setting a parameter*: you receive some data (need to define an API), assert the input data (e.g., are units/types are correct), derive the model parameters (e.g., a coordinate transformation or PSF simulations), validate that the outcome is consistent, and then review everything before accepting it. We have used this expression for a long time and and most documents (e.g., the concept document), my suggestion is not to change it.

File: README.md [hunk 6]
```diff
@@ -2,53 +2,92 @@
...
-```
+1. Receive: data or parameter update through an API.
+2. Assert: verify that input data or parameters are in correct units, formats, allowed ranges, etc.
+3. Derive: derive simulation model parameter from received data (not applicable to all input data).
+4. Validate: validate updated simulation model parameter (e.g., by comparison with measurements).
+5. Review: review of updates to simulation model parameter(s) and validation steps (by expert; in some cases automatized).
 
-Install dependencies using mamba or conda:
-```
+Acceptance Workflows:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:17]: These are workflows which are part of the acceptance process: someone or machine has to look at the validation / data assertion output from the setting workflow; if successful we update the model database, and everything has to be documented. I like the bullet points, as it amplifies the different steps.

File: README.md [hunk 7]
```diff
@@ -2,53 +2,92 @@
...
+
+1. Feedback: provide feedback on validation and review (successful / unsuccessful validation).
+2. Update DB: update simulation model database with new parameter (successful validation).
+3. Document: document all derivation and validation steps.
+
+The implementation in *simtools-workflows* consist of the following main components:
+
+- **workflows** encoded in [Common Workflow Language (CWL)](https://www.commonwl.org/) (see [./workflows](./workflows) in this repository)
+- **parameter schemas** defining all input data and simulation model parameters (see [./schemas](./schemas) in this repository)
+- **simulation tools and software** consisting of [simtools](https://github.com/gammasim/simtools) and the simulation software (e.g., [CORSIKA](https://www.iap.kit.edu/corsika/) and [sim_telarray](https://www.mpi-hd.mpg.de/hfm/~bernlohr/sim_telarray/)).
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:27]: Good point, I've changed the sentence so that it is clearer that simtools are not part of the workflows. For containers: I seem them now as part of simtools, so not really necessary to mention them again.

File: README.md [hunk 8]
```diff
@@ -2,53 +2,92 @@
...
+
+## Workflows and tools
+
+CWL workflows consists of
+
+- *Tools* are called by *steps* in a *Workflow* and are doing one single task. In most cases, tools are calling an application of [simtools](https://github.com/gammasim/simtools) including all required configuration parameters (e.g., [workflows/tools/derive_array_elements_coordinates.cwl](workflows/tools/derive_array_elements_coordinates.cwl)).
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:33]: Hmm, not sure - my feeling is that it is the same, just with adding `: they`. Let's not get hold up on wording.

File: README.md [hunk 9]
```diff
@@ -2,53 +2,92 @@
...
+- *Workflows* connect *tools*, and allow to execute the steps discussed above (receive, assert, derive, validate, ...), see.e.g, workflow/DeriveArrayElementCoordinates.cwl](./workflows/DeriveArrayElementCoordinates.cwl).
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:34]: Same comment as above, this too wordy.

File: README.md [hunk 10]
```diff
@@ -2,53 +2,92 @@
...
+```bash
 cwltool DeriveArrayElementCoordinates.cwl ../tests/resources/test_derive_array_elements_coordinates.yml
 ```
-The workflow steps are:
+
+The workflow steps executed are:
+
 - assert input data
 - convert coordinates
 - assert derive parameter values
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:61]: Yes.

File: README.md [hunk 11]
```diff
@@ -2,53 +2,92 @@
...
+mamba activate simtools-workflows-dev
 ```
 
-## Running examples
+All examples require docker to be installed and running.
+Tools are using the [simtools-prod:latest](https://github.com/gammasim/simtools/pkgs/container/simtools-prod) docker image.
+
+### Example workflow
 
 The following command line tool converts telescope coordinates from UTM to CORSIKA coordinates.
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:51]: The coordinates are not defined in the yml file, but the configuration is in that file. I've added a sentence to make this clearer.

File: README.md [hunk 12]
```diff
@@ -2,53 +2,92 @@
...
-(see the log files produced)
+Expected output:
 
-## CWL validation
+- telescope position file in CORSIKA coordinates (ecsv format)
+- metadata files describing each workflow step (yml format)
+- log files (stdout and stderr; ascii format)
 
-Use `cwltool --validate file_name.cwl` to check a workflow or command line tool for valid CWL.
+### CWL validation
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:69]: If changed the sentence to say  `check...for valid CWL code`. This is like a linter and is not really connected the the assert steps (assert does more than validate).

File: README.md [hunk 13]
```diff
@@ -2,53 +2,92 @@
...
 
-## Graphs
+Use `cwltool --validate file_name.cwl` to check a workflow file or command line tool to valid CWL code (this is also done by the CI).
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:71]: It is to check for `valid CWL code`. It doesn't check the workflow, it does check that keywords are used correctly and all required keywords are in the cwl file.

File: README.md [hunk 14]
```diff
@@ -2,53 +2,92 @@
...
+
+### Workflow Graphs
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:73]: As discussed, this is part of the code organisation / validation, not of the workflow validation.

File: README.md [hunk 15]
```diff
@@ -2,53 +2,92 @@
...
 
 Prepare a workflow graph, e.g.:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:75]: see comment above.

File: README.md [hunk 16]
```diff
@@ -2,53 +2,92 @@
...
-```
+
+```bash
 cwltool --print-dot DeriveArrayElementCoordinates.cwl | dot -Tsvg > DeriveArrayElementCoordinates.cwl.svg
 ```
 
 Alternatively, use https://view.commonwl.org/ , e.g., see [this example](https://view.commonwl.org/workflows/github.com/gammasim/workflows/blob/prototype-DeriveArrayElementCoordinates/workflows/DeriveArrayElementCoordinates.cwl).
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:81]: Not sure what this comments applies to - `use it to what` means?

File: README.md [hunk 17]
```diff
@@ -2,53 +2,92 @@
...
 
+## Parameter schema files
+
+The parameter schema files are used to describe all input data and simulation model parameters and can be found in [./schemas](./schemas) in this repository.
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:85]: Yes. Every model parameter has a schema. This is for the parameter definition, not for the values. Values are stored in the database, but not the definition.

File: schemas/MST_mirror_2f_measurements.schema.yml [hunk 18]
```diff
@@ -1,44 +1,39 @@
+%YAML 1.2
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply schemas/MST_mirror_2f_measurements.schema.yml:1]: Yes, this is a schema.

File: schemas/MST_mirror_2f_measurements.schema.yml (hunk 1 above)
Comment (GernotMaier) [reply schemas/MST_mirror_2f_measurements.schema.yml:35]: This is a first example, let's move the discussion on column names somewhere else. This is what we have used in the whole simtelarray/model validation until now and is a first example.

File: schemas/README.md (hunk 2 above)
Comment (GernotMaier) [reply schemas/README.md:72]: It means that astropy units can use these units. How and which functions are called is not so important in this context - isn't `compatible` saying exactly this?

File: schemas/README.md (hunk 3 above)
Comment (GernotMaier) [reply schemas/README.md:97]: But astropy understands "km" or "dimensionless_unscaled" (see discussion #9). I really don't see a reason to make it look more complicated than it has to.

The comment that units a required to be listed (this is not a choice; json-schema requires it) is mentioned further down.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply README.md:7]: Ok. So even if we use dirac for the other workflows, the dirac-kind of workflows are not going to be written here?

**Comment by GernotMaier** [comment]
Thanks for the review - I think I've addressed all comments.

Most importantly, I've re-organized the README.md so that it is clearer what part is for users and what is for developers.

Let me know if this is good.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply README.md:9]: Ok! I had not understood it before but it is clear now, thanks.

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply README.md:92]: Ah, sorry, probably I placed the comment on the wrong place.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply README.md:9]: I think a good solution is to put "The" before "Setting", then it is clearer that we are talking about a type of workflows and about the procedure to set a workflow.

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply README.md:84]: Ok, maybe you show me one of the graphs next week again so that I get a better understanding of what they are. I tried to open the link with the example you provided in the text and got an error: `Error: The workflow could not be retrieved from the Git repository using the details given`

**Review by VictorBarbosaMartins — APPROVED** [review]
Hi @GernotMaier. Thanks for updating the docs and for the explanations before! It is clearer now. I just left two conversations open with suggestions but I am fine with all the changes.

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:84]: Fixed the path to the graph.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply README.md:9]: Let's not discuss if we have to add 'The' or not - for me this is an additional word without information.

**Comment by GernotMaier** [comment]
@VictorBarbosaMartins - thanks for the review! I've fixed the path and will merge now.
